File: lib/chapter_edge_guard.dart
/// 章节边界二次确认换章（对齐原生 PagesManager isEndL/isEndR）。
enum ChapterEdgeOutcome { confirmNeeded, openChapter, atEnd }

class ChapterEdgeGuard {
  bool _hintNext = false;
  bool _hintPrev = false;

  ChapterEdgeOutcome onEdge(bool goNext, {required bool hasAdjacent}) {
    if (!hasAdjacent) {
      return ChapterEdgeOutcome.atEnd;
    }
    if (goNext ? _hintNext : _hintPrev) {
      if (goNext) {
        _hintNext = false;
      } else {
        _hintPrev = false;
      }
      return ChapterEdgeOutcome.openChapter;
    }
    if (goNext) {
      _hintNext = true;
    } else {
      _hintPrev = true;
    }
    return ChapterEdgeOutcome.confirmNeeded;
  }

  void clear() {
    _hintNext = false;
    _hintPrev = false;
  }

  void clearSide(bool goNext) {
    if (goNext) {
      _hintNext = false;
    } else {
      _hintPrev = false;
    }
  }
}

/// 一次 pointer 手势内只放行一次越界计数，避免同一次拖动的连续 overscroll 重复触发。
class EdgeGestureGate {
  bool _consumed = true;

  void beginGesture() {
    _consumed = false;
  }

  bool allow() {
    if (_consumed) return false;
    _consumed = true;
    return true;
  }
}

/// 越界 overscroll 是否计为一次章节边界手势。
/// 必须先确认方向对应边界，再消耗 [gate]，避免方向抖动白白吞掉本手势配额。
bool tryAcceptEdgeOverscroll({
  required double overscroll,
  required bool Function(bool towardEnd) atChapterEdge,
  required EdgeGestureGate gate,
  double minAbs = 6,
}) {
  if (overscroll.abs() < minAbs) return false;
  final towardEnd = overscroll > 0;
  if (!atChapterEdge(towardEnd)) return false;
  return gate.allow();
}

File: app/src/main/java/top/fumiama/copymangaweb/tool/ChapterEdgeGuard.kt
package top.fumiama.copymangaweb.tool

/** 章节边界二次确认结果（对齐 Flutter ChapterEdgeOutcome）。 */
enum class ChapterEdgeOutcome {
    CONFIRM_NEEDED,
    OPEN_CHAPTER,
    AT_END,
}

/**
 * 章节边界二次确认状态机（对齐原生旧 isEndL/isEndR 与 Flutter ChapterEdgeGuard）。
 * 纯逻辑，无 Android 依赖，可供 JVM 单测。
 */
class ChapterEdgeGuard {
    private var hintNext = false
    private var hintPrev = false

    fun onEdge(goNext: Boolean, hasAdjacent: Boolean): ChapterEdgeOutcome {
        if (!hasAdjacent) return ChapterEdgeOutcome.AT_END
        if (if (goNext) hintNext else hintPrev) {
            if (goNext) hintNext = false else hintPrev = false
            return ChapterEdgeOutcome.OPEN_CHAPTER
        }
        if (goNext) hintNext = true else hintPrev = true
        return ChapterEdgeOutcome.CONFIRM_NEEDED
    }

    fun clear() {
        hintNext = false
        hintPrev = false
    }

    fun clearSide(goNext: Boolean) {
        if (goNext) hintNext = false else hintPrev = false
    }
}
